mlir/lib/Dialect/Arith/Transforms/IntRangeOptimizations.cpp

```diff
@@ -185,25 +176,11 @@ struct IntRangeOptimizationsPass
     if (failed(solver.initializeAndRun(op)))
       return signalPassFailure();
 
-    DataFlowListener listener(solver);
-
-    RewritePatternSet patterns(ctx);
-    populateIntRangeOptimizationsPatterns(patterns, solver);
-
-    GreedyRewriteConfig config;
-    config.listener = &listener;
-
-    if (failed(applyPatternsAndFoldGreedily(op, std::move(patterns), config)))
-      signalPassFailure();
+    doRewrites(solver, ctx, op->getRegions());
   }
 };
 } // namespace
 
-void mlir::arith::populateIntRangeOptimizationsPatterns(
```

Comment:
Downstream we have a pass which uses our custom analysis, derived from `IntegerRangeAnalysis` with those patterns, but I guess, we can just make `doRewrites` function public (I can do it later in separate PR).

Reply:
There was a part of me that wanted to make these rewrites into patterns (using the match any operation setup) but there didn't seem to be a nice way to do it, especially since we're targeting block arguments.

(This PR came from the fact that our downstream has had a copy-pasted and hacked-on version of the test pass from when I first made the range analysis, and once I saw someone had made a `-int-range-optimizations`, I figured that code should migrate in to it)

Reply:
If you'd like these various optimizations structured as patterns and think it's safe to do so, I can do that.

(and I'm realizing the block argument constant replacements can be don from the parent op)

Reply:
I think, we can make pattern, which matches on any op and updates both op results and block args inside the op. this will also allow to get rid of manual worklist management and just delegate it to greedy rewriter driver.

Reply:
And done